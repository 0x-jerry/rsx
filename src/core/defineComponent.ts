import { isFn } from '@0x-jerry/utils'
import type { Merge } from 'type-fest'
import type { ToMaybeRef } from '.'
import type { DefineProps } from './props'

export interface PropOption<T = any> {
  type?: T
  default?: T | (() => T)
}

export type PropsType<T = any> = Record<string, T | PropOption<T>>

export interface ComponentOption<Props extends PropsType = {}> {
  props?: Props
  toRefs?: boolean
}

interface CommonProps {
  class?: any
  [key: string]: any
}

export type FunctionalComponent<P extends PropsType = {}> = (
  props: Merge<P, CommonProps>,
  children?: any[],
) => JSX.Element

/**
 * without option
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * with option
 * @param opt
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P>,
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * implement
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P> | FunctionalComponent,
  impl?: FunctionalComponent,
): FunctionalComponent {
  if (isFn(opt)) return opt

  // todo check props type
  return impl!
}

export const dc = defineComponent
