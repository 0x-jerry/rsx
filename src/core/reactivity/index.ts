export { computed, reactive, ref } from '@vue/reactivity'
export * from './effect'
export { nextTick } from './scheduler'
export * from './types'
export * from './utils'
export * from './watch'
