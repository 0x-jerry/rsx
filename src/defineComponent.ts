import type { EmptyObject } from '@0x-jerry/utils'
import type { DefineProps } from './props'
import type { IsStartWithCapitalizedLetter, MaybeRef } from './types'

export type PropsType<T = any> = Record<string, T>

type ToPropRefs<T extends {}> = {
  /**
   * Start with capitalized letter means this prop is not used as reactive prop
   */
  [key in keyof T]: key extends IsStartWithCapitalizedLetter<key>
    ? T[key]
    : MaybeRef<T[key]>
}

/**
 * @private
 */
export type FunctionalComponent<P extends PropsType = any> = (
  props: P,
  children?: any[],
) => unknown

type _NoInferProps<P extends EmptyObject> = DefineProps<P> & Record<string, any>

export type ExposedFunctionalComponent<P extends PropsType = any> =
  FunctionalComponent<ToPropRefs<P>>

/**
 * Auto calculate props from generic type P
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: NoInfer<FunctionalComponent<_NoInferProps<P>>>,
): ExposedFunctionalComponent<P>

/**
 * Infer props from impl parameter
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<P>,
): ExposedFunctionalComponent<P>

/**
 * implement
 */
export function defineComponent(impl?: any): any {
  // todo check props type
  return impl
}

export const dc = defineComponent
