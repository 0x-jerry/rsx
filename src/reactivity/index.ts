export { computed, reactive, ref } from '@vue/reactivity'
export * from './bindingRef'
export * from './effect'
export { nextTick } from './scheduler'
export * from './types'
export * from './watch'
