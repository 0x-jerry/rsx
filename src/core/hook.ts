import type { Fn } from '@0x-jerry/utils'
import type { WatchCallback, WatchEffect, WatchSource } from '@vue/reactivity'
import {
  type DNodeContext,
  DNodeContextEventName,
  getCurrentContext,
} from './context'
import { type WatchHandle, type WatchOptions, watch } from './reactivity'

export function unmount(ctx: DNodeContext) {
  ctx.emit(DNodeContextEventName.beforeUnmount)

  ctx.el?.remove()

  ctx.children?.forEach((child) => unmount(child))

  ctx.emit(DNodeContextEventName.unmounted)
}

export function mount(ctx: DNodeContext) {
  ctx.emit(DNodeContextEventName.beforeMount)

  ctx.children?.forEach((item) => mount(item))

  ctx.emit(DNodeContextEventName.mounted)
}

export function useContext() {
  const ctx = getCurrentContext()

  if (!ctx) {
    throw new Error('This should only used inside functional component.')
  }

  return ctx
}

export function onBeforeMount(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.beforeMount, fn)
}

export function onMounted(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.mounted, fn)
}

export function onBeforeUnmount(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.beforeUnmount, fn)
}

export function onUnmounted(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.unmounted, fn)
}

export function useWatch(
  getter: WatchSource | WatchSource[] | WatchEffect,
  fn: WatchCallback,
  opt?: WatchOptions,
): WatchHandle {
  const stop = watch(getter, fn, opt)

  onUnmounted(stop)

  return stop
}
