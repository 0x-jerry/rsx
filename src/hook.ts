import type { EmptyObject, Fn } from '@0x-jerry/utils'
import {
  readonly,
  type WatchCallback,
  type WatchEffect,
  type WatchSource,
} from '@vue/reactivity'
import {
  type DNodeContext,
  DNodeContextEventName,
  type DNodeEventMap,
  getCurrentContext,
} from './context'
import { isRef, type WatchHandle, type WatchOptions, watch } from './reactivity'

export function useContext() {
  const ctx = getCurrentContext()

  if (!ctx) {
    throw new Error('This should only used inside functional component.')
  }

  return ctx
}

function createContextHook(name: keyof DNodeEventMap) {
  return (fn: Fn, ctx?: DNodeContext) => {
    ctx ||= useContext()

    ctx.on(name, fn)
  }
}

export const onBeforeMount = createContextHook(
  DNodeContextEventName.beforeMount,
)
export const onMounted = createContextHook(DNodeContextEventName.mounted)

export const onBeforeUnmount = createContextHook(
  DNodeContextEventName.beforeUnmount,
)
export const onUnmounted = createContextHook(DNodeContextEventName.unmounted)

export function useWatch(
  getter: WatchSource | WatchSource[] | WatchEffect,
  fn: WatchCallback,
  opt?: WatchOptions,
): WatchHandle {
  const stopHandle = watch(getter, fn, opt)

  onUnmounted(stopHandle)

  return stopHandle
}

export type InjectionKey<T> = (string | symbol) & { _: T }

export function provide<T = unknown>(
  key: string | symbol | InjectionKey<T>,
  value: T,
) {
  const ctx = useContext()
  ctx.ex ||= {}
  ctx.ex[key] = value
}

export function inject<T>(
  key: string | symbol | InjectionKey<T>,
): T | undefined {
  const ctx = useContext()

  let node = ctx.parent

  while (node) {
    if (node.ex?.[key]) {
      return node.ex[key] as T
    }

    node = node.parent
  }
}

export function useRawProps() {
  const ctx = useContext()

  return ctx._node?.props || {}
}

export function useExpose<T extends EmptyObject>(exposed: T) {
  const ref = useRawProps().ref

  if (isRef(ref)) {
    ref.value = readonly(exposed)
  }
}

export function useRawChildren() {
  return useContext()._node?.children || []
}
