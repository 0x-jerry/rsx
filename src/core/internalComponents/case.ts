import type { JsonPrimitive, Optional } from '@0x-jerry/utils'
import type { DNodeContext } from '../context'
import { defineComponent, type FunctionalComponent } from '../defineComponent'
import { createDynamicNode, dispatchMovedEvent } from '../dynamicNode'
import { mount, onBeforeMount, unmount, useWatch } from '../hook'
import { createComponentInstance, transformProps } from '../jsx'
import { insertBefore } from '../nodeOp'
import { $, computed } from '../reactivity'

export interface CaseComponentProps {
  condition: JsonPrimitive
  cases?: Record<string, Optional<FunctionalComponent>>
}

export const VCase = defineComponent<CaseComponentProps>((props) => {
  const el = createDynamicNode()

  const caseKey = computed(() => String(props.condition))

  let renderedCtxNode: Optional<DNodeContext> = null

  useWatch(caseKey, updateCase)

  onBeforeMount(updateCase)

  el.addEventListener('moved', () => {
    if (renderedCtxNode?.el) {
      insertBefore(el, renderedCtxNode.el)
      dispatchMovedEvent(renderedCtxNode.el)
    }
  })

  return el

  function updateCase() {
    if (renderedCtxNode) {
      unmount(renderedCtxNode)
      renderedCtxNode = null
    }

    renderedCtxNode = rebuildChild()
  }

  function rebuildChild() {
    const render = props.cases?.[caseKey.value]
    if (!render) {
      return
    }

    const newChild = createComponentInstance(render)
    if (newChild.el) {
      insertBefore(el, newChild.el)
    }

    mount(newChild)

    return newChild
  }
})

export interface IfComponentProps {
  condition: JsonPrimitive
  truthy?: FunctionalComponent
  falsy?: FunctionalComponent
}

export const VIf = defineComponent<IfComponentProps>((props) => {
  const _props = transformProps(VIf, {
    condition: $(() => !!props.condition),
    cases: {
      true: props.truthy,
      false: props.falsy,
    },
  })

  return VCase(_props)
})
