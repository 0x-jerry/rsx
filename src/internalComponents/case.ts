import type { JsonPrimitive, Optional } from '@0x-jerry/utils'
import { createAnchorNode } from '@/nodes/AnchorNode'
import { mount, unmount } from '@/nodes/lifeCycle'
import { asyncWatcherScheduler } from '@/reactivity/scheduler'
import { defineComponentName } from '@/test'
import { runWithContext } from '../context'
import { defineComponent, type FunctionalComponent } from '../defineComponent'
import { onBeforeMount, useContext, useWatch } from '../hook'
import { insertBefore } from '../nodeOp'
import { type ComponentNode, createComponentNode } from '../nodes/ComponentNode'
import { normalizeProps } from '../props'
import { $, computed } from '../reactivity'

export interface CaseItemComponentProps<T> {
  value: T
}

export type CaseItemComponent<T> = FunctionalComponent<
  CaseItemComponentProps<T>
>

export interface CaseItem<T> {
  where: (value: T) => boolean
  render: CaseItemComponent<T>
}

export interface CaseComponentProps<T = unknown> {
  condition: T
  cases?:
    | Record<string, Optional<CaseItemComponent<NoInfer<T>>>>
    | CaseItem<NoInfer<T>>[]
}

export const VCase = defineComponent(<T>(props: CaseComponentProps<T>) => {
  const ctx = useContext()

  const anchorNode = createAnchorNode('case')

  const ChildComponent = computed(() => {
    if (Array.isArray(props.cases)) {
      return props.cases.find((n) => n.where(props.condition))?.render
    }

    return props.cases?.[String(props.condition)]
  })

  let renderedNode: Optional<ComponentNode>
  let initialized = false

  const rebuildChildren = () => runWithContext(updateCase, ctx)

  useWatch(ChildComponent, rebuildChildren, {
    scheduler: asyncWatcherScheduler,
  })

  onBeforeMount(() => {
    rebuildChildren()
    initialized = true
  })

  return anchorNode

  function updateCase() {
    if (renderedNode) {
      unmount(renderedNode)
      renderedNode = null
    }

    renderedNode = rebuildChild()
  }

  function rebuildChild() {
    const Component = ChildComponent.value

    if (!Component) {
      return
    }

    const childNode = createComponentNode(
      Component,
      { value: $(() => props.condition) },
      [],
    )

    childNode.initialize()

    if (childNode.el) {
      insertBefore(anchorNode.el!, childNode.el)
    }

    if (initialized) {
      mount(childNode)
    }

    anchorNode.children = [childNode]

    return childNode
  }
})

defineComponentName(VCase, 'VCase')

export interface IfComponentProps {
  condition: JsonPrimitive
  truthy?: FunctionalComponent
  falsy?: FunctionalComponent
}

export const VIf = defineComponent<IfComponentProps>((props) => {
  const _props = normalizeProps(VIf, {
    condition: $(() => !!props.condition),
    cases: {
      true: props.truthy,
      false: props.falsy,
    },
  })

  return VCase(_props as CaseComponentProps)
})

defineComponentName(VIf, 'VIf')
