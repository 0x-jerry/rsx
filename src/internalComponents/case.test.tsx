import { dc } from '../defineComponent'
import { onMounted } from '../hook'
import { $, nextTick, ref } from '../reactivity'
import { contextToJson, defineComponentName, mountTestApp } from '../test'
import { VCase, VIf } from './case'

describe('VCase', () => {
  it('non-reactivity data', async () => {
    const App = dc(() => {
      const value = 0

      return (
        <VCase
          condition={value}
          cases={{
            '0': () => <div class="item">0</div>,
            '1': () => <div class="item">1</div>,
          }}
        />
      )
    })

    const el = mountTestApp(App)

    const getContent = () =>
      el
        .querySelectorAll('.item')
        .values()
        .map((item) => item.textContent)
        .toArray()

    expect(getContent()).eql(['0'])
  })

  it('reactivity data', async () => {
    const App = dc(() => {
      const value = ref(0)

      onMounted(() => {
        value.value = 1
      })

      return (
        <VCase
          condition={value}
          cases={{
            '0': () => <div class="item">0</div>,
            '1': () => <div class="item">1</div>,
          }}
        />
      )
    })

    const el = mountTestApp(App)

    const getContent = () =>
      el
        .querySelectorAll('.item')
        .values()
        .map((item) => item.textContent)
        .toArray()

    expect(getContent()).eql(['1'])
  })

  it('where case condition', async () => {
    const value = ref(7)

    const App = dc(() => {
      return (
        <VCase
          condition={value}
          cases={[
            {
              where: (value) => value > 10,
              render: (props) => <p>{$(() => props.value)} 1</p>,
            },
            {
              where: (value) => 5 > value,
              render: (props) => <p>{$(() => props.value)} 2</p>,
            },
            {
              where: () => true,
              render: (props) => <p>{$(() => props.value)} 3</p>,
            },
          ]}
        />
      )
    })

    const el = mountTestApp(App)

    const getContent = () => el.querySelector('p')?.textContent

    expect(getContent()).eql('7 3')

    value.value = 11
    await nextTick()
    expect(getContent()).eql('11 1')

    value.value = 3
    await nextTick()
    expect(getContent()).eql('3 2')
  })
})

describe('VIf', () => {
  it('non-reactivity data', async () => {
    const App = dc(() => {
      const value = 0

      return (
        <VIf
          condition={value}
          truthy={() => <div class="item">1</div>}
          falsy={() => <div class="item">0</div>}
        />
      )
    })

    const el = mountTestApp(App)

    const getContent = () =>
      el
        .querySelectorAll('.item')
        .values()
        .map((item) => item.textContent)
        .toArray()

    expect(getContent()).eql(['0'])
  })

  it('reactivity data', async () => {
    const App = dc(() => {
      const value = ref(0)

      onMounted(() => {
        value.value = 1
      })

      return (
        <VIf
          condition={value}
          truthy={() => <div class="item">1</div>}
          falsy={() => <div class="item">0</div>}
        />
      )
    })

    const el = mountTestApp(App)

    const getContent = () =>
      el
        .querySelectorAll('.item')
        .values()
        .map((item) => item.textContent)
        .toArray()

    expect(getContent()).eql(['1'])
  })
})

describe('case context tree', () => {
  it('static value', () => {
    const A = dc((_, children) => (
      <div class="A">
        <span>a</span>
        {children}
      </div>
    ))

    defineComponentName(A, 'A')

    const B = dc((_, children) => (
      <div class="B">
        <span>b</span>
        {children}
      </div>
    ))

    defineComponentName(B, 'B')

    const App = dc(() => {
      const value = 0

      return (
        <VCase
          condition={value}
          cases={{
            '0': () => <A>0</A>,
            '1': () => <B>1</B>,
          }}
        />
      )
    })

    defineComponentName(App, 'App')

    const root = mountTestApp(App)

    expect(root).toMatchSnapshot('html')

    const ctxTree = contextToJson(root._)

    expect(ctxTree).toMatchSnapshot('ctx tree')
  })

  it('reactive value', async () => {
    const A = dc((_, children) => (
      <div class="A">
        <span>a</span>
        {children}
      </div>
    ))

    defineComponentName(A, 'A')

    const B = dc((_, children) => (
      <div class="B">
        <span>b</span>
        {children}
      </div>
    ))

    defineComponentName(B, 'B')

    const App = dc(() => {
      const value = ref(0)

      onMounted(() => {
        value.value = 1
      })

      return (
        <VCase
          condition={value}
          cases={{
            '0': () => <A>0</A>,
            '1': () => <B>1</B>,
          }}
        />
      )
    })

    defineComponentName(App, 'App')

    const root = mountTestApp(App)

    expect(root).toMatchSnapshot('html')

    const ctxTree = contextToJson(root._)

    expect(ctxTree).toMatchSnapshot('ctx tree')
  })
})
