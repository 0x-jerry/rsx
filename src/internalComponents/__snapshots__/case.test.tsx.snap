// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`case context tree > reactive value > ctx tree 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "B",
            },
          ],
          "name": "1",
        },
      ],
      "name": "VCase",
    },
  ],
  "name": "Root",
}
`;

exports[`case context tree > reactive value > html 1`] = `
<div
  id="app"
>
  <div
    class="B"
  >
    <span>
      b
    </span>
    1
  </div>
  <!--case-->
</div>
`;

exports[`case context tree > static value > ctx tree 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "A",
            },
          ],
          "name": "0",
        },
      ],
      "name": "VCase",
    },
  ],
  "name": "Root",
}
`;

exports[`case context tree > static value > html 1`] = `
<div
  id="app"
>
  <div
    class="A"
  >
    <span>
      a
    </span>
    0
  </div>
  <!--case-->
</div>
`;
